docs/09-progressive-web-app.md

```diff
@@ -92,3 +92,143 @@ Once installed, the app will appear as a standalone app in its own window with n
+
+```cs
+async Task RequestNotificationSubscriptionAsync()
+{
+    var subscription = await JSRuntime.InvokeAsync<NotificationSubscription>("blazorPushNotifications.requestSubscription");
+    if (subscription != null)
+    {
+        await HttpClient.PutJsonAsync<object>("notifications/subscribe", subscription);
+    }
+}
+```
+
+This code invokes a JavaScript function that you'll find in `BlazingPizza.ComponentsLibrary/wwwroot/pushNotifications.js`. The JavaScript code there calls the `pushManager.subscribe` API and returns the results to .NET.
+
+If the user agrees to receive notifications, this code sends the data to your server where the tokens are stored in your database for later use.
+
+To try this out, start placing an order and go to the checkout screen. You should see a request:
+
+![image](https://user-images.githubusercontent.com/1101362/66354176-eed8eb80-e95b-11e9-9799-b4eba6410971.png)
+
+Choose *Allow* and check in the browser dev console that it didn't cause any errors. If you want, set a breakpoint on the server in `NotificationsController`'s `Subscribe` action method, and run with debugging. You should be able to see the incoming data from the browser, which includes an endpoint URL as well as some cryptographic tokens.
+
+Once you've either allowed or blocked notifications for a given site, your browser won't ask you again. If you need to reset things for further testing, if you're using Chrome or Edge beta, you can click the "information" icon to the left of the address bar, and change *Notifications* back to *Ask (default)* as in this screenshot:
+
+![image](https://user-images.githubusercontent.com/1101362/66354317-58f19080-e95c-11e9-8c24-dfa2d19b45f6.png)
+
+### Sending a notification
+
+Now you have subscriptions, you can send notifications. This involves performing some complex cryptographic operations on your server in order to protect the data in transit. Thankfully the bulk of the complexity is handled for us by a third-party NuGet package.
+
+To get started, in your `BlazingPizza.Server` project, reference the NuGet package `WebPush`. The following instructions are based on using version `1.0.11`.
+
+Next, open `OrdersController`. Have a look at the `TrackAndSendNotificationsAsync` method. This simulates a sequence of delivery steps once each order is placed, and calls `SendNotificationAsync` which isn't yet fully implemented.
+
+We'll now update `SendNotificationAsync` to actually dispatch notifications using the subscription you captured earlier for the order's user. The following code makes uses of `WebPush` APIs for dispatching the notification:
+
+```cs
+private static async Task SendNotificationAsync(Order order, NotificationSubscription subscription, string message)
+{
+    // For a real application, generate your own
+    var publicKey = "BLC8GOevpcpjQiLkO7JmVClQjycvTCYWm6Cq_a7wJZlstGTVZvwGFFHMYfXt6Njyvgx_GlXJeo5cSiZ1y4JOx1o";
```

Comment:
Does the WebPush package have docs on how to do this? It might be nice to link to that here.

Reply:
Oh nevermind, I see it's below.